### ISSUE TITLE
 Training Stuck with Unexpected Number of Epochs (fine-tuning /starcoder)

### ISSUE THREAD

**Issue by tclxmeng-jia** [bigcode-project/octopack#13]
### Description

I'm encountering an issue while fine-tuning `starcoder` using the provided script. The training seems to be stuck, and I'm getting an unexpected number of epochs. Here's the log information:

***** Running training *****
Num examples = 4,000
Num Epochs = 9,223,372,036,854,775,807
Instantaneous batch size per device = 1
Total train batch size (w. parallel, distributed & accumulation) = 4
Gradient Accumulation steps = 4
Total optimization steps = 1,000
Number of trainable parameters = 35,553,280

UserWarning: MatMul8bitLt: inputs will be cast from torch.float32 to float16 during quantization
warnings.warn(f"MatMul8bitLt: inputs will be cast from {A.dtype} to float16 during quantization")
{'loss': 1.1266, 'learning_rate': 0.0001, 'epoch': 0.0}

***** Running Evaluation *****
Num examples: Unknown
Batch size = 1

It appears that the training process is stuck and the number of epochs is unexpectedly high. I'm running the script with the following parameters:

```bash
python finetune.py \
    --model_path="bigcode/starcoder" \
    --dataset_name="ArmelR/guanaco-commits" \
    --seq_length 2048 \
    --max_steps 1000 \
    --batch_size 1 \
    --input_column_name="content" \
    --gradient_accumulation_steps 4 \
    --learning_rate 5e-4 \
    --lr_scheduler_type="cosine"\
    --log_freq 1 \
    --eval_freq 1 \
    --num_warmup_steps 5 \
    --save_freq 5 \
    --weight_decay 0.05 \
    --output_dir="/root/autodl-fs/checkpoints"\
    --no_fp16 \
    --streaming
```

I'm not sure why the number of epochs is unexpectedly high and why the training process is not progressing as expected. Could you please help me understand what might be causing this issue?

**Comment by ArmelRandy** [comment]
Hi @tclxmeng-jia. About the unexpectedly high number of epochs, don't worry about this because it happens when transformers is not able to precisely compute the number of epochs. It is the case here because you did load the dataset in streaming. Trainer does not know in advance how many samples there are in the dataset, so it can not compute the precise number of epochs.  Moreover, the `streaming` argument was thought for large datasets, and in this case, the validation set is generated on by taking the first `size_valid_set` samples. You should not use this because I think you use a pretty small dataset that is already divided into `train` and `test`. Also, do you have a GPU? Launching your code with `python` is not advisable here. And finally, I am not sure that this dataset has a column `content`.

Can you check all these things?

**Comment by tclxmeng-jia** [comment]
@ArmelRandy Thank you for your guidance. I will make the necessary adjustments and conduct a thorough review based on your suggestions. It appears I overlooked utilizing the GPU, and I appreciate your reminder. Once again, thank you for your assistance.

**Comment by tclxmeng-jia** [comment]
@ArmelRandy What are the recommended hardware requirements for fine-tuning? I'm currently using a single A40 GPU with 48GB of memory, and the training speed is noticeably slow. Would it be advisable to consider adding more GPUs to potentially improve the training performance? My training dataset includes code snippets in the 'content' column, which are quite lengthy. I'm uncertain whether the issue is related to my dataset or hardware limitations. Below is the training progress I've collected over the course of two days:

```js
[
  {'loss': 0.7188, 'learning_rate': 0.0001, 'epoch': 0.0},
  {'loss': 0.8503, 'learning_rate': 0.0002, 'epoch': 0.0},
  {'loss': 1.1605, 'learning_rate': 0.0003, 'epoch': 0.0}, 
  {'loss': 1.1378, 'learning_rate': 0.0004, 'epoch': 0.0}, 
  {'loss': 1.0755, 'learning_rate': 0.0005, 'epoch': 0.01}, 
  {'loss': 1.0278, 'learning_rate': 0.0004999987538703314, 'epoch': 0.01}, 
  {'loss': 0.9878, 'learning_rate': 0.000499995015493748, 'epoch': 0.01}, 
  {'loss': 0.7929, 'learning_rate': 0.000499988784907518, 'epoch': 0.01},
   ....,
    {'eval_loss': 0.849025547504425, 'eval_runtime': 6721.9799, 'eval_samples_per_second': 0.652,'eval_steps_per_second': 0.217, 'epoch': 0.02},
    ....,
    {'loss': 1.0147, 'learning_rate': 0.0004997557981487559, 'epoch': 0.02},
106 {'eval_loss': 0.8489658832550049, 'eval_runtime': 6725.0896, 'eval_samples_per_second': 0.652, 'eval_steps_per_second': 0.217, 'epoch': 0.02},
107 {'loss': 1.1401, 'learning_rate': 0.0004997196729961109, 'epoch': 0.02}
]
```
I appreciate any insights you can provide. Thank you for your time.
